File: Java/046_Permutations.java
public class Solution {
    public List<List<Integer>> permute(int[] nums) {
        List<List<Integer>> res = new ArrayList<List<Integer>>();
        List<Integer> tmp = new ArrayList<>();
        
        for(int e : nums) tmp.add(e);
        res.add(tmp);   
        int[] next = Arrays.copyOf(nums, nums.length);
        nextPermutation(next);
        while(!Arrays.equals(next, nums)) {
        	List<Integer> temp = new ArrayList<>();
        	for(int e : next) temp.add(e);
        	res.add(temp);
        	nextPermutation(next);
        }
        
        return res;
        
    }
	
	public void nextPermutation(int[] nums) {
		if(nums.length <= 1 || nums == null) return;
		int i = nums.length-2;
		while(i >= 0 && nums[i] >= nums[i+1]) i--;
		if(i >= 0) {
			int j = nums.length-1;
			while(nums[j] <= nums[i]) j--;
			swap(nums, i, j);
		}
		reverse(nums, i+1, nums.length-1);
	}
	
	private void swap(int[] nums, int i, int j) {
		int tmp = nums[i];
		nums[i] = nums[j];
		nums[j] = tmp;
	}
	
	private void reverse(int[] nums, int from, int to) {
		while(from < to)
			swap(nums, from++, to--);
	}
}

File: Cpp/046_Permutations.cpp
class Solution {
public:
	vector<vector<int>> permute(vector<int>& nums) {
		vector<vector<int>> res;
		res.push_back(nums);
		vector<int> next(nums);
		nextPermutation(next);
		while(next != nums) {
			vector<int> tmp(next);
			res.push_back(tmp);
			nextPermutation(next);
		}
		return res;
	}

	void nextPermutation(vector<int>& nums) {
		if (nums.size() <= 1) return;
		int i = nums.size()-2;
		while (i >= 0 && nums[i] >= nums[i+1]) i--;
		if (i >= 0) {
			int j = nums.size()-1;
			while (nums[j] <= nums[i]) j--;
			swap(nums[i], nums[j]);
		}
		reverse(nums, i+1, nums.size()-1);
	}
	void reverse(vector<int> &nums, int from, int to) {
		while(from < to) swap(nums[from++], nums[to--]);
	}

};

//回溯法（速度慢）
class Solution {
public:
    vector<vector<int>> permute(vector<int>& nums) {
		vector<vector<int>> res;
		vector<int> temp;
		backTrack(res, temp, nums);
		return res;
	}
	void backTrack(vector<vector<int>> &res, vector<int> &temp, vector<int> &nums) {
		if(temp.size() == nums.size()) {
			vector<int> list(temp);
			res.push_back(list);
		} else {
			for (int i = 0; i < nums.size(); i++) {
				if (find(temp.begin(), temp.end(), nums[i]) != temp.end())
					continue;
				temp.push_back(nums[i]);
				backTrack(res, temp, nums);
				temp.pop_back();
			}
		}
	}
};
